kamu-core/src/infra/repos/object_repository_s3.rs

```diff
@@ -187,6 +192,29 @@ where
         Ok(Box::new(stream))
     }
 
+    async fn get_download_url(&self, _prefix_url: &Url, hash: &Multihash) -> Result<(Url, Option<DateTime<Utc>>), GetError> {
+        let key = self.get_key(hash);
+        let get_object_request = GetObjectRequest {
+            bucket: self.bucket.clone(),
+            key,
+            ..GetObjectRequest::default()
+        };
+
+        let provider = ChainProvider::new();
+        let credentials = provider.credentials().await.unwrap();
+
+        let validity_period_seconds: u64 = 3600;
+        let options = PreSignedRequestOption {
+            expires_in: std::time::Duration::from_secs(validity_period_seconds),
+        };
+        
+        let presigned_url = get_object_request.get_presigned_url(&(Region::default()), &credentials, &options);
+        match Url::parse(presigned_url.as_str()) {
+            Ok(url) => Ok((url, Some(Utc::now() + chrono::Duration::seconds(validity_period_seconds as i64)))),
+            Err(e) => Err(GetError::Internal(e.int_err()))
+        }
+    }
```

Comment:
This feels misplaced to me in the `ObjectRepository` interface... but I need to think about this more to propose an alternative.

Reply:
It feels the same way too, and this is related to `prefix_url` as well, which is ignored here, as you can see, and only used in local FS. I thought of a Visitor pattern, but the implementation depends on the internal fields a lot.

Reply:
Some poorly structured thoughts:

Imagine we're doing a "pull" via `https://my-odf-server.com/:accountName/:datasetName/pull` - when transferring data I see following possible modes of operation:
- **Proxied** - data is served through simple transfer protocol
  - URL is determined as `reverse_url(/blocks/:hash)`
  - e.g. `https://my-odf-server.com/:accountName/:datasetName/blocks/:hash`
- **Delegated** - downloads/uploads go directly to storage
  - e.g. `https://s3.us-west-2...aws.com/datasetsBucket/:datasetID/blocks/:hash`
- **In-Band** - data is streamed via same websocket (we're not implementing this yet, but useful to keep in mind)

Some ODF servers can only choose to support proxied mode, some storage back-ends may not support delegation, some clients may prefer only proxied/in-band mode (e.g. some firewall concerns).

Now how this all relates to `ObjectRepository` - should it be the one issuing these URLs or do we need a completely different interface for this...

The platform-specific implementation of the `ObjectRepository` can be pretty complex:
- It may utilize some cache / index to read blocks really fast
  - e.g. Redis / Elastic, or holding last 100 blocks of datasets in memory...
- Writes and cache misses will have to use a URL of the **internal** storage, e.g. `s3://datasets.kamu.dev/:datasetID/blocks/`
- Which in most cases will be different from public storage URLs for downloads, e.g. `https://some-cdn-host.com/bucketname/:datasetID/blocks/:hash`


Putting everything together I think:
- I don't see a reason to move `get_download_url()` out of `ObjectRepository` yet
- I think `prefix_url` param should be removed though
- Instead we should be configuring `ObjectRepository` with both
  - internal URL to use for normal reads/writes
  - external URL for issuing download/upload URLs to the clients


I suggest to change the signature to:
```rust
trait ObjectRepository {
  async fn get_download_url(&self, hash: &Multihash, opts: &DownloadOpts)
    -> Result<GetDownloadUrlResult, GetDownloadUrlError>;
}

struct DownloadOpts {
  expiration: Option<TimeDelta>,
}

struct GetDownloadUrlResult {
  url,
  expires_at,
}

enum GetDownloadUrlError {
  NotSupported,
  ...
}
```

This way if a specific storage back-end does not support delegated downloads/uploads - the "smart" handler can fall back onto proxied / "simple" protocol. Note that this will require some sort of URL reversal `reverse_url("relative/url")` - which Axum may have support for.

Reply:
Changed the signature of `get_download_url` as described above. 
Eliminated `prefix_url` argument from all operations.
Shifted resolution of the base URL to dataset resolver - passing external base URL as an argument.
Local FS repository optionally can store a base URL, which is used to reply on download links